a10_octavia/controller/worker/flows/a10_load_balancer_flows.py

```diff
@@ -505,3 +517,59 @@ def get_delete_lb_vrid_subflow(self):
             requires=[a10constants.VRID, a10constants.DELETE_VRID]))
 
         return delete_lb_vrid_subflow
+
+    def _get_cascade_delete_listeners_flow(self, lb):
+        """Sets up an internal delete flow
+        Because task flow doesn't support loops we store each listener
+        we want to delete in the store part and then rebind
+        :param lb: load balancer
+        :return: (flow, store) -- flow for the deletion and store with all
+                 the listeners stored properly
+        """
+        listeners_delete_flow = unordered_flow.Flow('listener_delete_flow')
+        store = {}
+        compute_flag = lb.amphorae
+        for listener in lb.listeners:
+            for l7policy in listener.l7policies:
+                l7policy_name = 'l7policy_' + l7policy.id
+                store[l7policy_name] = l7policy
+                listeners_delete_flow.add(
```

Comment:
This will lead to an unintended side effect where an attempt is made to delete an L7 policy before the listener as it's being added to an unordered flow.

Instead, create an UnorderedFlow to start. On each iteration, create a linear flow and add the listener delete flow to it then add the L7 policy delete flow to it.

For 3 listeners this would look like

```
UnorderedFlow( LinearFlow( ListenerDeleteFlow() -> L7PolicyDeleteFlow() ),
               LinearFlow( ListenerDeleteFlow() -> L7PolicyDeleteFlow() )
               LinearFlow( ListenerDeleteFlow() -> L7PolicyDeleteFlow() )
````


This model can be extended to the member , health monitor, pool side of things too.

Technically this entire construct could be handled in one function. No real need to separate Listener from pool. My guess was that it was done to increase read ability.

Reply:
Done the changes as discussed over teams